server/store/subscriptions.go

```diff
@@ -0,0 +1,253 @@
+		ApprovalStatus:                   subscription.ApprovalStatus,
+		BuildStatusName:                  subscription.BuildStatusName,
+		StageNameValue:                   subscription.StageNameValue,
+		ReleasePipelineName:              subscription.ReleasePipelineName,
+		ReleaseStatusName:                subscription.ReleaseStatusName,
+		ApprovalTypeName:                 subscription.ApprovalTypeName,
+		ApprovalStatusName:               subscription.ApprovalStatusName,
+		RunPipeline:                      subscription.RunPipeline,
+		RunPipelineName:                  subscription.RunPipelineName,
+		RunStageName:                     subscription.RunStageName,
+		RunEnvironmentName:               subscription.RunEnvironmentName,
+		RunStageNameID:                   subscription.RunStageNameID,
+		RunStageStateID:                  subscription.RunStageStateID,
+		RunStageStateIDName:              subscription.RunStageStateIDName,
+		RunStageResultID:                 subscription.RunStageResultID,
+		RunStateID:                       subscription.RunStateID,
+		RunStateIDName:                   subscription.RunStateIDName,
+		RunResultID:                      subscription.RunResultID,
+	}
+	subscriptionList.ByMattermostUserID[userID][subscriptionKey] = subscriptionListValue
+}
+
+func (s *Store) GetSubscriptionList() (*SubscriptionList, error) {
+	key := GetSubscriptionListMapKey()
+	initialBytes, appErr := s.Load(key)
+	if appErr != nil {
+		return nil, errors.New(constants.GetSubscriptionListError)
+	}
+
+	subscriptions, err := SubscriptionListFromJSON(initialBytes)
+	if err != nil {
+		return nil, errors.New(constants.GetSubscriptionListError)
+	}
+
+	return subscriptions, nil
+}
+
+func (s *Store) GetAllSubscriptions(userID string) ([]*serializers.SubscriptionDetails, error) {
+	subscriptions, err := s.GetSubscriptionList()
+	if err != nil {
+		return nil, err
+	}
+
+	var subscriptionList []*serializers.SubscriptionDetails
+	if userID == "" {
+		for mmUserID := range subscriptions.ByMattermostUserID {
+			for _, subscription := range subscriptions.ByMattermostUserID[mmUserID] {
+				subscription := subscription // we need to do this to prevent implicit memory aliasing in for loop
+				subscriptionList = append(subscriptionList, &subscription)
+			}
+		}
+	} else {
+		for _, subscription := range subscriptions.ByMattermostUserID[userID] {
+			subscription := subscription
+			subscriptionList = append(subscriptionList, &subscription)
+		}
+	}
+
+	return subscriptionList, nil
+}
+
+func deleteSubscriptionAtomicModify(subscription *serializers.SubscriptionDetails, initialBytes []byte) ([]byte, error) {
+	subscriptionList, err := SubscriptionListFromJSON(initialBytes)
+	if err != nil {
+		return nil, err
+	}
+	subscriptionKey := GetSubscriptionKey(subscription.MattermostUserID,
+		subscription.ProjectName,
+		subscription.ChannelID,
+		subscription.EventType,
+		subscription.Repository,
+		subscription.TargetBranch,
```

Comment:
We're also duplicating information about the assigned channel (channel name and channel type), which could change at any time.

I'm also thinking we should support a `Subscription Title` field, so an admin can identify the purpose of a given subscription.

Reply:
Actually, there is no option to edit an existing subscription
one user can just view another user's subscriptions if that is created in a channel he has access to

also, MM users who are admins on an Azure DevOps project can only create/delete a subscription as this is a limitation by Azure DevOps
so, on MM only project admins can create subscriptions and delete each other's subscriptions while non-admin users can just view subscriptions in MM's channel that he has access to

Reply:
> Yes, the current logic for forming a subscription id/key with md5 may cause an issue if fields are added/dropped, we can look on having unique IDs

@avas27JTG This is something you want to do up front though. Otherwise you have to do perform a migration to migrate the data to use IDs later, and it's best to avoid these migrations if you can because of the implications of running multiple instances of the plugin simultaneously.

It's okay to not have subscriptions be editable on this release, but they should definitely have unique IDs to support editing the subscriptions in a future release. We can use them for deleting subscriptions now as well. I think we should remove the md5 complication altogether.

> For 1 we can not change an assigned channel for a subscription as we cannot edit it.

Right, I meant the channel name and/or channel type can change at any time

Reply:
Fixed, now we are using the subscription id returned by the Azure DevOps for storing in the KV store